### PR TITLE
React to SameSite addition to CookieOptions

### PR DESCRIPTION
**Pull request by JunTaoLuo**
#127 We should be setting the default to strict?

### CODE REVIEW

File: src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs [hunk 1]
```diff
@@ -69,9 +69,13 @@ public void SaveCookieToken(HttpContext httpContext, string token)
...
             Debug.Assert(token != null);
 
-            var options = new CookieOptions();
-            options.HttpOnly = true;
-            options.Domain = _options.CookieDomain;
+            var options = new CookieOptions
+            {
+                HttpOnly = true,
+                Domain = _options.CookieDomain,
+                SameSite = SameSiteEnforcementMode.Strict
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs]: I am not familiar with this option, Let's have a conversation with @blowdart - we need to understand what should be the default and **if** this should be configurable.

Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs]: Spoke offline, I think this should remain as strict as we expect the cookie to be created by requests from the same site as subsequent requests. There's no need to make this configurable.

**Comment by JunTaoLuo** [comment]
Updated based on @blowdart's recommendation to make this option configurable. This is ready to merge.

**Comment by rynowak** [comment]
If we're going to make this configurable we should attach an `Action<CookieOptions>` to `AntiforgeryOptions` and be done. Then users are empowered to do whatever they want and we don't end up duplicating all of `CookieOptions` on `AntiforgeryOptions`.

File: src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs [hunk 2]
```diff
@@ -54,10 +54,11 @@ public string CookieName
...
 
         /// <summary>
-        /// The domain set on the cookie. By default its <c>null</c> which results in the "domain" attribute not being
-        /// set.
+        /// Configures the <see cref="CookieOptions"/> of the antiforgery cookies. Without additional configuration, the 
+        /// default values antiforgery cookie options are true for <see cref="CookieOptions.HttpOnly"/>, null for
+        /// <see cref="CookieOptions.Domain"/> and <see cref="SameSiteMode.Strict"/> for <see cref="CookieOptions.SameSite"/>.
         /// </summary>
-        public string CookieDomain { get; set; }
+        public Action<CookieOptions> ConfigureCookieOptions { get; set; }
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs]: This is now a breaking change that requires an announcement.

File: src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs [hunk 3]
```diff
@@ -80,6 +84,8 @@ public void SaveCookieToken(HttpContext httpContext, string token)
             }
             SetCookiePath(httpContext, options);
 
+            _options.ConfigureCookieOptions?.Invoke(options);
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs]: I call the new ConfigureCookieOptions last after everything else is set. If we don't want users to be able to override Secure or Path on the CookieOptions, we can move this up, though I don't see why we might want that.

File: src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs]: Get @Eilon to sign off on the breaking change, I don't get to approve these

Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs]: Added him as a reviewer and will speak with him about the breaking change. Can you review the rest of the PR?

File: src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs (hunk 3 above)
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs]: Do we also want to configure the cookie path and secure in the delegate?

Comment (rynowak) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs]: Personally I think allowing you to set both is fine. RequireSSL is also enforced on the validation side, so if you mess it up, you'll just get a 400.

We should double check with @blowdart

File: test/Microsoft.AspNetCore.Antiforgery.Test/Internal/DefaultAntiforgeryTokenStoreTest.cs [hunk 4]
```diff
@@ -364,7 +364,7 @@ public void SaveCookieToken_NonNullAntiforgeryOptionsCookieDomain_UsesOptionsCoo
                 .Returns("/index.html");
             var options = new AntiforgeryOptions();
             options.CookieName = _cookieName;
-            options.CookieDomain = expectedCookieDomain;
+            options.ConfigureCookieOptions = cookieOptions => cookieOptions.Domain = expectedCookieDomain;
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Antiforgery.Test/Internal/DefaultAntiforgeryTokenStoreTest.cs]: Change the name of this test to reflect the change.

File: src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs (hunk 3 above)
Comment (blowdart) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs]: Consistency is good, but folks are already using the old way. I assume you don't want to remove the existing path, just add a way to set it here to?

Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs]: Yea, not changing the existing behaviour. Calling the configure delegate here allows the user to overwrite what we set but as @rynowak mentioned, they will be doing so at their own risk.

File: test/Microsoft.AspNetCore.Antiforgery.Test/Internal/DefaultAntiforgeryTokenStoreTest.cs (hunk 4 above)
Comment (JunTaoLuo) [inline test/Microsoft.AspNetCore.Antiforgery.Test/Internal/DefaultAntiforgeryTokenStoreTest.cs]: Done

File: src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs [hunk 5]
```diff
@@ -54,12 +54,21 @@ public string CookieName
         public PathString? CookiePath { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs]: This should be obsolete too, it's redundant with ConfigureCookieOptions

File: src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs [hunk 6]
```diff
@@ -69,9 +69,13 @@ public void SaveCookieToken(HttpContext httpContext, string token)
             Debug.Assert(httpContext != null);
             Debug.Assert(token != null);
 
-            var options = new CookieOptions();
-            options.HttpOnly = true;
-            options.Domain = _options.CookieDomain;
+            var options = new CookieOptions
+            {
+                HttpOnly = true,
+                Domain = null,
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs]: still use the obsolete domain property

File: src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs [hunk 7]
```diff
@@ -69,9 +69,13 @@ public void SaveCookieToken(HttpContext httpContext, string token)
             Debug.Assert(httpContext != null);
             Debug.Assert(token != null);
 
-            var options = new CookieOptions();
-            options.HttpOnly = true;
-            options.Domain = _options.CookieDomain;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs]: We still need to use `_options.CookieDomain` - it's not being removed.

Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs]: ah yes

**Comment by JunTaoLuo** [comment]
:up::date:

File: src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs [hunk 8]
```diff
@@ -69,9 +69,13 @@ public void SaveCookieToken(HttpContext httpContext, string token)
...
+            var options = new CookieOptions
+            {
+                HttpOnly = true,
+                Domain = null,
+                SameSite = SameSiteMode.Strict
+            };
+
             // Note: don't use "newCookie.Secure = _options.RequireSSL;" since the default
             // value of newCookie.Secure is populated out of band.
             if (_options.RequireSsl)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs]: This is misleading, remove it.

**Comment by JunTaoLuo** [comment]
As per discussion with @Tratcher, we will probably adopt this approach for other components that duplicate settings on CookieOptions, potentially updating things like Session. Will file an issue in HttpAbstractions to track.

**Pull request merged**